library/std/src/io/copy.rs

```diff
@@ -0,0 +1,450 @@
+            CopyParams(fd_to_meta(self), Some(self.as_raw_fd()))
+        }
+    }
+
+    impl CopyWrite for StdoutLock<'_> {
+        fn properties(&self) -> CopyParams {
+            CopyParams(fd_to_meta(self), Some(self.as_raw_fd()))
+        }
+    }
+
+    impl CopyWrite for StderrLock<'_> {
+        fn properties(&self) -> CopyParams {
+            CopyParams(fd_to_meta(self), Some(self.as_raw_fd()))
+        }
+    }
+
+    impl<T: CopyRead> CopyRead for Take<T> {
+        fn drain_to<W: Write>(&mut self, writer: &mut W, outer_limit: u64) -> Result<u64> {
+            let local_limit = self.limit();
+            let combined_limit = min(outer_limit, local_limit);
+            let bytes_drained = self.get_mut().drain_to(writer, combined_limit)?;
+            // update limit since read() was bypassed
+            self.set_limit(local_limit - bytes_drained);
+
+            Ok(bytes_drained)
+        }
+
+        fn min_limit(&self) -> u64 {
+            min(Take::limit(self), self.get_ref().min_limit())
+        }
+
+        fn properties(&self) -> CopyParams {
+            self.get_ref().properties()
+        }
+    }
+
+    impl<T: CopyRead> CopyRead for BufReader<T> {
+        fn drain_to<W: Write>(&mut self, writer: &mut W, outer_limit: u64) -> Result<u64> {
+            let buf = self.buffer();
+            let buf = &buf[0..min(buf.len(), outer_limit.try_into().unwrap_or(usize::MAX))];
+            let bytes = buf.len();
+            writer.write_all(buf)?;
+            self.consume(bytes);
+
+            let remaining = outer_limit - bytes as u64;
+
+            // in case of nested bufreaders we also need to drain the ones closer to the source
+            let inner_bytes = self.get_mut().drain_to(writer, remaining)?;
+
+            Ok(bytes as u64 + inner_bytes)
+        }
+
+        fn min_limit(&self) -> u64 {
+            self.get_ref().min_limit()
+        }
+
+        fn properties(&self) -> CopyParams {
+            self.get_ref().properties()
+        }
+    }
+
+    impl<T: CopyWrite> CopyWrite for BufWriter<T> {
+        fn properties(&self) -> CopyParams {
+            self.get_ref().properties()
+        }
+    }
+
+    fn fd_to_meta<T: AsRawFd>(fd: &T) -> FdMeta {
+        let fd = fd.as_raw_fd();
+        let file: ManuallyDrop<File> = ManuallyDrop::new(unsafe { File::from_raw_fd(fd) });
+        match file.metadata() {
+            Ok(meta) => FdMeta::Metadata(meta),
+            Err(_) => FdMeta::None,
+        }
+    }
+
+    unsafe fn fd_as_file(fd: RawFd) -> ManuallyDrop<File> {
+        ManuallyDrop::new(File::from_raw_fd(fd))
+    }
+
+    #[cfg(test)]
+    mod tests {
+
+        use crate::io::{BufRead, BufReader, BufWriter, Read, Result, Seek, SeekFrom, Write};
+
+        #[test]
+        fn copy_specialization() -> Result<()> {
+            let path = crate::env::temp_dir();
+            let source_path = path.join("copy-spec.source");
+            let sink_path = path.join("copy-spec.sink");
```

Comment:
Does std have some policy about tests polluting the filesystem? there doesn't seem to be any utility for temporary files and libc doesn't even have a wrapper for `memfd_create` :(